### PR TITLE
Handle Layered Component Trees for per sprite-layer drawdepths

### LINKED ISSUE #4803: Sprite rendering should be based around layers not sprites
Consider that sprite rendering, at least from the cpu end, is a similar problem space to physics + dynamictrees:

Ideally:
- Each drawdepth is on its own dynamictree. We have far fewer of those than collision layers so we would probably only have 20 or 30 and not 100.
- When we get sprites to draw we can sort per dynamictree and not globally which is much faster and means we can dispatch sprites for drawing much faster.
- Cleaner separation when we want to do overlays in-between drawdepths.
- Per-layer drawdepths is useful in some rare cases where you really don't want 2 entities.

### PR DESCRIPTION
**Pull request by AndrewL97**
About the PR
Create a new LayeredComponentTreeComponent that allows LayeredComponentTreeSystems to handle a dictionary/layers of binary trees, rather than a single flat one.

Why / Balance
Fixes #4803
Allows us to have sprites that draw layers on different drawdepths. So entities can pass "between" another entity's layers. Think gates, doorways, etc.

Technical details
Added a LayeredComponentTreeSystem that lets us handle splitting Component Trees into Layers.

Changed SpriteTreeSystem to use the new layered system.

Added handling to the Sprite component and rendering pipeline to fetch and draw based on drawdepth/layer.
The sprite and index lists are now dictionaries, so the individual sorts don't need to handle drawDepth anymore.

the Layers config on sprite components now lets you define a drawDepth. If we don't define one, we default to the previous layer's drawDepth, or parent Sprite's if no other layers exist.
No changes have been made to the existing entities, so there shouldn't actually be any visible changes from this PR.

Changelog

tweak: Added handling for drawing sprite layers on a drawdepth separate to their parent sprite.

### CODE REVIEW

**Comment by AndrewL97** [comment]
@metalgearsloth this is my first foray into rendering _and_ SS14, so please let me know if I have botched standards, etc.
Very much open to feedback <3

File: Robust.Shared/Utility/CollectionExtensions.cs [hunk 1]
```diff
@@ -40,7 +40,8 @@ public static List<T> ShallowClone<T>(this List<T> self)
 
         public static bool TryGetValue<T>(this IList<T> list, int index, out T value)
         {
-            if (list.Count > index)
+
+            if (list.Count > index && index >= 0)
             {
                 value = list[index];
```
Comment (metalgearsloth) [inline Robust.Shared/Utility/CollectionExtensions.cs:63]: You should avoid autoformatting stuff like this as it just makes the review harder to read.

Comment (AndrewL97) [reply Robust.Shared/Utility/CollectionExtensions.cs:63]: Fixed!

**Comment by metalgearsloth** [comment]
I think I like this more than https://github.com/space-wizards/RobustToolbox/pull/4905 and would rather pull emo's ideas in but I have a content backlog of stuff I really need to go through first.

**Comment by AndrewL97** [comment]
Hey @metalgearsloth  - is this something you guys are still keen for? There are conflicts for me to fix now, but I'll hold off unless this is something we want :)

**Comment by metalgearsloth** [comment]
Electro left some comments here on emo's PR https://github.com/space-wizards/RobustToolbox/pull/4905#issuecomment-2066861749

Moving sprite rendering to another entity's responsibility would probably greatly simplify the logic involved here.

**Comment by AndrewL97** [comment]
Easy peasy, I'll close this one and have a squizz

**Comment by PJB3005** [comment]
FWIW I've been meaning to review this for ages, I just haven't gotten around to it yet.

**Comment by AndrewL97** [comment]
@PJB3005  
> FWIW I've been meaning to review this for ages, I just haven't gotten around to it yet.

All G - I haven't looked at this since my last commit anyway, so I'll restart with Electro's comment in mind
Not 100% sure how the entity sprite layer combination stuff will work in practice, but I'll annoy people in Discord about it

**Pull request closed without merging**